console/v1/types_console_sample.go

```diff
@@ -0,0 +1,193 @@
+
+	// tags like "Java", "Quarkus", etc.
+	// Tags are optional and can be used by the user to find samples in the samples catalog.
+	// They will be also displayed on the samples details page.
+	// +optional
+	// +listType=set
+	// +kubebuilder:validation:MaxItems:=10
+	Tags []string `json:"tags"`
+
+	// The actual sample source, like an external git repository or container image.
+	// +kubebuilder:validation:Required
+	Source ConsoleSampleSource `json:"source"`
+}
+
+// ConsoleSampleSourceType is an enumeration of the supported sample types.
+// Unsupported samples types will be ignored in the web console.
+// +kubebuilder:validation:Enum:=GitImport;ContainerImport
+type ConsoleSampleSourceType string
+
+const (
+	// A sample that let the user import code from a git repository.
+	GitImport ConsoleSampleSourceType = "GitImport"
+	// A sample that let the user import a container image.
+	ContainerImport ConsoleSampleSourceType = "ContainerImport"
+)
+
+// ConsoleSampleSource is the actual sample definition and can hold different sample types.
+// Unsupported sample types will be ignored in the web console.
+// +union
+// +kubebuilder:validation:XValidation:rule="self.type == 'GitImport' ? has(self.gitImport) : !has(self.gitImport)",message="source.gitImport is required when source.type is GitImport, and forbidden otherwise"
+// +kubebuilder:validation:XValidation:rule="self.type == 'ContainerImport' ? has(self.containerImport) : !has(self.containerImport)",message="source.containerImport is required when source.type is ContainerImport, and forbidden otherwise"
+type ConsoleSampleSource struct {
+	// type of the sample, currently supported: "GitImport";"ContainerImport"
+	// +unionDiscriminator
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Enum:="GitImport";"ContainerImport"
+	Type ConsoleSampleSourceType `json:"type"`
+
+	// gitimport let the user import code from a git repository.
+	// +unionMember
+	// +optional
+	GitImport *ConsoleSampleGitImportSource `json:"gitImport,omitempty"`
+
+	// containerimport let the user import a container image.
+	// +unionMember
+	// +optional
+	ContainerImport *ConsoleSampleContainerImportSource `json:"containerImport,omitempty"`
+}
+
+// ConsoleSampleGitImportSource let the user import code from a public git repository.
+type ConsoleSampleGitImportSource struct {
+	// url of the Git repository. Only public HTTPS repositories on GitHub, GitLab
+	// and Bitbucket are currently supported:
+	//
+	//   - https://github.com/<org>/<repository>
+	//   - https://gitlab.com/<org>/<repository>
+	//   - https://bitbucket.org/<org>/<repository>
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=100
+	// +kubebuilder:validation:Pattern=`^https:\/\/(github.com|gitlab.com|bitbucket.org)\/[a-zA-Z0-9-]+\/[a-zA-Z0-9-]+(.git)?$`
+	URL string `json:"url"`
+	// revision to clone a specific branch, tag or commit SHA, etc.
```

Comment:
Is there a maximum length we can apply to this one?

Reply:
Added MaxLength=256